truss/truss_handle.py

```diff
@@ -397,11 +397,50 @@ def training_docker_build_setup(self, build_dir: Optional[Path] = None):
 
     def add_python_requirement(self, python_requirement: str):
         """Add a python requirement to truss model's config."""
-        self._update_config(
-            lambda conf: replace(
-                conf, requirements=[*conf.requirements, python_requirement]
+
+        # Parse the added python requirements
+        requirement_name = self._python_req_name(python_requirement)
+        requirement_version = self._python_req_version(python_requirement)
+
+        # Parse the existing requirements
+        requirements = self._existing_req_names(self._spec.config.requirements)
+
+        # If we already have the requirement among the existing requirements
+        # but version needs to be updated
+        if requirement_name in map(lambda reqs: reqs.name, requirements):
+            _ = [
+                self._existing_req_version_update(req, requirement_version)
+                for req in requirements
+                if req.name == requirement_name
+            ]
+            self._update_config(lambda conf: replace(conf, requirements=[]))
+
+        # If the requirement is totally new
+        if python_requirement not in requirements:
+            self._update_config(
+                lambda conf: replace(
+                    conf, requirements=[*conf.requirements, python_requirement]
+                )
             )
+
+    @staticmethod
+    def _python_req_name(python_requirement: str) -> str:
+        return pkg_resources.Requirement.parse(python_requirement).name
+
+    @staticmethod
+    def _python_req_version(python_requirement: str) -> str:
+        return list(pkg_resources.Requirement.parse(python_requirement).specs[0])[1]
+
+    @staticmethod
+    def _existing_req_names(req_names: list[str]) -> list[str]:
+        return list(pkg_resources.parse_requirements(req_names))
+
+    @staticmethod
+    def _existing_req_version_update(requirement: str, new_version: str) -> list[str]:
+        list(requirement.specs[0])[1].replace(
+            list(requirement.specs[0])[1], new_version
         )
+        return requirement
```

Comment:
Can you also move this function outside of the class? Any where above the class definition is fine

```suggestion
def _python_req_name(python_requirements:str):
        return pkg_resources.Requirement.parse(python_requirement).name
```

Reply:
I did the modifications and added tests. I couldnt run the tests because of a missing file, but let me know please if they are not okay

Thank you for your help and everything @pankajroark and @bolasim